pos-print/src/main/java/com/clouway/pos/print/adapter/http/PrintServiceV2.kt

```diff
@@ -0,0 +1,97 @@
+import com.google.sitebricks.headless.Service
+import com.google.sitebricks.http.Get
+import com.google.sitebricks.http.Post
+import org.slf4j.LoggerFactory
+import javax.inject.Inject
+
+/**
+ * @author Tsvetozar Bonev (tsvetozar.bonev@clouway.com)
+ */
+@Service
+@At("/v2/receipts/req/print")
+class PrintServiceV2 @Inject constructor(private var repository: ReceiptRepository,
+                                         private var queue: PrintQueue) {
+  private val logger = LoggerFactory.getLogger(PrintServiceV2::class.java)
+
+  @Post
+  fun printReceipt(request: Request): Reply<*> {
+    return try {
+      val dto = request.read(ReceiptDTO::class.java).`as`(GsonTransport::class.java)
+
+      val receiptRequest = ReceiptRequest(dto.receipt, dto.sourceIp, dto.operatorId, dto.fiscal)
+
+      val requestId = repository.register(receiptRequest)
+
+      queue.queue(receiptRequest.copy(requestId = requestId))
+
+      logger.info("Receipt with request id $requestId queued for printing")
+      Reply.with(requestId).status(202)
+    } catch (ex: ReceiptAlreadyRegisteredException) {
+      logger.error("Receipt is already in queue")
+      Reply.saying<Unit>().badRequest()
+    }
+  }
+
+  @Get
+  @At("/status/:requestId")
+  fun getReceiptStatus(@Named("requestId") requestId: String): Reply<*> {
+    return try {
+      val receiptStatus = repository.getStatus(requestId)
+      logger.info("Receipt status returned as ${receiptStatus.name}")
+      Reply.with(receiptStatus).ok()
+    } catch (ex: ReceiptNotRegisteredException) {
+      logger.error("Receipt with request id $requestId was  not found")
+      Reply.saying<Unit>().notFound()
+    }
+  }
+
+  @Get
+  @At("/receipts/:limit")
+  fun getLatestReceipts(@Named("limit") limit: Int): Reply<*> {
+    val receipts = repository.getLatest(limit)
+
+    val receiptStatuses = mutableMapOf<String, String>()
+
+    receipts.forEach{
+      receiptStatuses[it.key] = repository.getStatus(it.key).name
+    }
+
+    val receiptList = mutableListOf<ReceiptWithStatusDTO>()
+
+    receiptStatuses.forEach{
+      receiptList.add(ReceiptWithStatusDTO(it.value, it.key, Gson().toJson(receipts[it.key]!!)))
+    }
+
+    return Reply.with(receiptList).ok()
+  }
+
+  @Post
+  @At("/requeue/:requestId")
+  fun requeueReceipt(@Named("requestId") requestId: String): Reply<*> {
+    return try {
+      val receiptPrintRequest = repository.getReceiptPrintRequest(requestId)
+      queue.queue(receiptPrintRequest)
+      Reply.saying<Unit>().ok()
+    } catch (e: ReceiptNotRegisteredException) {
+      logger.error("Receipt with request id $requestId was not found")
+      Reply.saying<Unit>().notFound()
+    }
+  }
+
+  internal data class ReceiptWithStatusDTO(val status: String, val requestId: String, val receiptJson: String)
+  internal data class ReceiptDTO(val sourceIp: String = "", val operatorId: String = "", val fiscal: Boolean = false, val receipt: Receipt = Receipt.newReceipt().build())
```

Comment:
tI think that it is better if this dto is PrintReceiptRequest(...)

it look like you are using the domain object as a dto : 
Receipt = Receipt.newReceipt().build()

you should create separate dto object that is used only for parsing the request , then you will have to make transformation from dto to the domain

Reply:
I implemented a ReceiptWithStatus object that is now stored in persistence, containing information about it's printing status, fiscal status, request id, source ip and its receipt object.

I don't think using the receipt domain object directly here will cause a problem, as the request contains a fully-fledged receipt object within itself already.